proposals/0540-jsem.rst

```diff
@@ -0,0 +1,342 @@
+        processes through the semaphore ⟨sem⟩.
+
+        Use of ``-jsem`` will override use of :ghc-flag:``-j[⟨n⟩]``,
+        and vice-versa.
+
+GHC Jobserver Protocol
+~~~~~~~~~~~~~~~~~~~~~~
+
+The GHC Jobserver Protocol allows a client to dynamically invoke many instances
+of GHC, while restricting all of those instances to use no more than
+<n> capabilities. This is achieved by coordination over a system semaphore
+(either a POSIX semaphore [6]_  in the case of Linux and Darwin, or a
+Win32 semaphore [7]_  in the case of Windows platforms).
+
+There are two kinds of participants in the GHC Jobserver protocol:
+
+- The *jobserver* creates a system semaphore with a certain number of
+  available tokens.
+
+  Each time the jobserver wants to spawn a new jobclient subprocess, it **must**
+  first acquire a single token from the semaphore, before spawning
+  the subprocess. This token **must** be released once the subprocess terminates.
+
+  Once work is finished, the jobserver **must** destroy the semaphore it created.
+
+- A *jobclient* is a subprocess spawned by the jobserver or another jobclient.
+
+  Each jobclient starts with one available token (its *implicit token*,
+  which was acquired by the parent which spawned it), and can request more
+  tokens within the Jobserver Protocol by waiting on the semaphore.
+
+  Each time a jobclient wants to spawn a new jobclient subprocess, it **must**
+  pass on a single token to the child jobclient. This token can either be the
+  jobclient's implicit token, or another token which the jobclient acquired
+  from the semaphore.
+
+  Each jobclient **must** release exactly as many tokens as it has acquired from
+  the semaphore (this does not include its implicit token).
+
+Implementation details
+~~~~~~~~~~~~~~~~~~~~~~
+
+There is a hidden cost in having a GHC program (such as ``ghc`` itself) run on
+fewer CPU cores than its capabilities: the stop-the-world cost of garbage
+collection becomes much more expensive. If we were to give GHC a fixed number
+``N`` of capabilities, then parallel garbage collections in that GHC would
+recruit ``N`` OS threads. Instead we propose that GHC will adjust its number
+of capabilities, via ``setNumCapabilities`` [8]_ , to the number of modules
+it is compiling.
+Note that GHC already calls ``setNumCapabilities`` with the argument to ``-j``.
+
+We have concerns that rapidly adjusting the number of capabilities may have
+adverse effects: we propose that GHC will rate limit the release of tokens so as
+to reduce this thrashing. The exact mechanism of rate limiting is unspecified
+here.
+
+We also believe that GHC invocations should not be too eager to release
+resources, so we propose a minimum delay between acquiring a resource and
+releasing a resource. This allows us to prioritise completing a single large
+package before moving on to other work. The exact mechanism through which this
+is achieved is also left unspecified in this proposal.
```

Comment:
I tried to clarify that one issue is that if you are compiling N packages at once you have N copies of the EPS so you will load interface files multiple times (once per GHC session) which will use more memory than if you are compiling one package with N capabilities. Therefore we are trying to skew the system so that we prefer compiling 1 with N rather than N with 1.

Reply:
Thanks, this makes sense.

The general idea is:
- Having several GHC running at the same time increases memory usage
- There is one GHC running by package
- If a package has enough parallelism to saturate the number of parallel job specified, it is therefore presumably better to use all the jobs on this packages: a single copy of GHC is running.
- To achieve this, you will make it so that a copy of GHC which already have jobs running is allowed to run new jobs, automatically having priority over other GHC instances.

Did I get this right?

Reply:
Yes that's right.